# Математические пакеты. Лабораторная работа №3

Выполнил студент группы 9381 Аухадиев Александр.

Вариант 3.

## Цели работы

Практическое применение библиотеки символьных вычислений Python (SymPy) для решения задач дискретной математики и математической логики.

## Задачи

1. Решить свой вариант диофантова уравнения.
2. Имея открытую часть ключа RSA найти соответствующую закрытую часть и расшифровать кодированное слово.
3. Записать в формате ДНФ и КНФ булеву функцию, заданную своим аналитическим выражением.



## Выполнение работы

### Задача 1. Диофантово уравнение

3825x + 4267y = 68

In [11]:
from sympy.solvers.diophantine import diophantine
from sympy import symbols
from sympy.solvers.diophantine.diophantine import diop_solve

In [8]:
x, y, z = symbols("x, y, z", integer = True)
diophantine(3825*x + 4267*y - 68)

{(251*t_0 - 116, 104 - 225*t_0)}

In [12]:
diop_solve(3825*x + 4267*y - 68)

(251*t_0 - 116, 104 - 225*t_0)

Ответ: x = -116, y = 104

### Задача 2. RSA кодирование

Русское слово из четырех букв закодировано при помощи алгоритма RSA открытым ключом (e = 5,m = 39). Шифрованное сообщение имеет вид (18; 22; 13; 35). Подберите закрытую часть ключа и прочитайте исходное слово. Буквам русского алфавита соответсвуют числа в диапазоне от 2 до 32 (исключены буквы “e” и “ъ”).

In [18]:
from sympy.crypto.crypto import rsa_private_key, decipher_rsa

In [20]:
code = [18, 22, 13, 35]
key = rsa_private_key(39, 5) 
answer = [decipher_rsa(i, key) for i in code]
print(answer)

[18, 16, 13, 29]


Ответ: РОЛЬ

### Задача 3. Преобразование булевой функции

((xy) v (y xor x)) xor ((zx) v z)

In [21]:
from sympy.logic.boolalg import to_cnf, to_dnf

In [24]:
DNF = to_dnf(((x&y)|(y^x))^((z&x)|z), simplify=True)
print(DNF)
CNF = to_cnf(((x&y)|(y^x))^((z&x)|z), simplify=True)
print(CNF)

(x & ~z) | (y & ~z) | (z & ~x & ~y)
(x | y | z) & (~x | ~z) & (~y | ~z)


## Выводы

В ходе выполнения лабораторной работы были изучены функции библиотеки символьных вычислений Python (SymPy) для решения задач дискретной математики и математической логики.